lib/engine-native/src/artifact.rs

```diff
@@ -593,7 +602,117 @@ impl Artifact for NativeArtifact {
+
+        let min_call_trampolines_pointer = self
+            .finished_function_call_trampolines
+            .values()
+            .map(|t| *t as usize)
+            .min()
+            .unwrap_or(0);
+        let min_dynamic_trampolines_pointer = self
+            .finished_dynamic_function_trampolines
+            .values()
+            .map(|t| **t as usize)
+            .min()
+            .unwrap_or(0);
+
+        let fp = self.finished_functions.clone();
+        let mut function_pointers = fp.into_iter().collect::<Vec<_>>();
+
+        // Sort the keys by the values in reverse order (function pointers)
+        // This way we can get the maximum function lengths (since functions can't collide in memory)
+        function_pointers.sort_by(|(_k1, v1), (_k2, v2)| v2.cmp(v1));
+        let mut function_pointers = function_pointers
+            .into_iter()
+            .map(|(index, function_pointer)| {
+                let fp = **function_pointer as usize;
+                let current_size_by_ptr = if prev_pointer != usize::MAX {
+                    // This assumes we never lay any functions bodies across the usize::MAX..nullptr
+                    // wrapping point.
+                    // Which is generally true on most OSes, but certainly doesn't have to be true.
+                    //
+                    // Further reading: https://lwn.net/Articles/342330/ \
+                    // "There is one little problem with that reasoning, though: NULL (zero) can
+                    // actually be a valid pointer address."
+                    prev_pointer - fp
+                } else {
+                    // We try to determine the size based on any of the trampolines that usually
+                    // go after that function (the   last defined in Wasm)
+                    if fp < min_call_trampolines_pointer {
+                        if min_call_trampolines_pointer < min_dynamic_trampolines_pointer
+                            || min_dynamic_trampolines_pointer == 0
+                        {
+                            min_call_trampolines_pointer - fp
+                        } else {
+                            min_dynamic_trampolines_pointer - fp
+                        }
+                    } else if fp < min_dynamic_trampolines_pointer {
+                        min_dynamic_trampolines_pointer - fp
+                    } else {
+                        // The trampoline pointers are before the function.
+                        // We can safely assume the function will be at least 16 bits of length.
+                        // This is a very hacky assumption, but it makes collisions work perfectly
+                        // Since DLOpen simlinks will always be > 16 bits of difference between
+                        // two different libraries for the same symbol.
+                        // Note: the minmum Mach-O file is 0x1000 bytes and the minimum ELF is 0x0060 bytes
+                        16
+                    }
```

Comment:
I tried to simplify this, also by moving it outside. It didn't work because it depends of the `fp` (function pointer) value.
I do think, however that comparing to usize::MAX is a bit hacky, so there might be some other cleaner way to do that.